app/backend/approaches/chatreadretrieveread.py

```diff
@@ -117,3 +122,16 @@ def get_chat_history_as_text(self, history: Sequence[dict[str, str]], include_la
             if len(history_text) > approx_max_tokens*4:
                 break    
         return history_text
+    
+    # Generate messages needed for chat Completion api
+    def get_messages_from_prompt(self, prompt: str) -> []:
+        messages = []
+        for line in prompt.splitlines():
+            if line.startswith("<|im_start|>"):
+                index = "<|im_start|>".__len__()
```

Comment:
I'm not sure why this is using the dunder method, I would typically use the built-in len().

Also, since line 131 will always yield the same answer, you can move it out of the loop, e.g.:

START = "<|im_start|>"
START_LEN = len(START)

messages = []
for line in prompt.splitlines():
    if line.startswith(START):
      role = line[START_LEN:]

That might be over-optimization though.

Reply:
Is it still necessary to use the im_start/im_end approach at all in this file? I wonder if we can move to a list for everything, or if thats incompatible with some aspect.

Reply:
Hi @srbalakr, thank you for raising PR.

I have few suggestions to reduce the number of tokens in the generated `messages` array:
1. Have a single `role` and `content` pair for each role.
2. Strip leading and trailing whitespaces in each line.
3. Ignore any empty lines if first or last line of content for each role.

In current implementation, each empty line will be added as `{'role': role, 'content': ''}` in `messages` array, e.g., for `system` role, it will be 11 tokens for `{'role': system, 'content': ''}`.

For a given sample prompt,
```
prompt = """\
<|im_start|>system
A

B


C
<|im_end|>
<|im_start|>user

Who are you?

<|im_end|>
<|im_start|>assistant

"""
````

**Current implementation output:**
```
[{'role': 'system', 'content': 'A'}, {'role': 'system', 'content': ''}, {'role': 'system', 'content': 'B'}, {'role': 'system', 'content': ''}, {'role': 'system', 'content': ''}, {'role': 'system', 'content': 'C'}, {'role': 'user', 'content': ''}, {'role': 'user', 'content': 'Who are you?'}, {'role': 'user', 'content': ''}, {'role': 'assistant', 'content': ''}]
```
Tokens = 133

**Expected output:**
```
[{'role': 'system', 'content': 'A\n\nB\n\n\nC'}, {'role': 'user', 'content': 'Who are you?'}]
```
Tokens = 44

Reply:
@pamelafox  @bharatsb thank you. I have refactored the code to use tiktok lib to estimate tokens. It makes code efficient and simpler. Help take a look.